### PR TITLE
Dev saas microsoft.saa s 2018 03 01 beta

### PR DESCRIPTION
**Pull request by diyakobo**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [x] new service onboarding 
      - [ ] new API version 
      - [ ] update existing version for new feature 
      - [x] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [x] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://msazure.visualstudio.com/One/_workitems/create/Scenario?templateId=d1c215ec-5d6a-4cb3-a193-95ddd0a1ba96&ownerId=85bb92fb-21fe-41ca-9309-47f2c5f57515). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by weidongxu-microsoft** [comment]
please check the checkbox.

File: specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json [hunk 1]
```diff
@@ -0,0 +1,866 @@
...
+          "Get saas operations": {
+            "$ref": "./examples/saasV1/SaaSGetOperations.json"
+          }
+        },
+        "x-ms-pageable": {
+          "nextLinkName": "nextLink"
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.SaaS/checkmoderneligibility": {
```
Comment (majastrz) [inline specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json:76]: >checkmoderneligibility [](start = 62, length = 22)

The swagger convention is lowerCaseCamelCase. Can you change it to `checkModernEligibility`?

File: specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json [hunk 2]
```diff
@@ -0,0 +1,866 @@
...
+        },
+        "x-ms-examples": {
+          "Get SaaS resource": {
+            "$ref": "./examples/saasV2/SaasGet.json"
+          }
+        }
+      }
+    },
+    "/providers/Microsoft.SaaS/saasresources": {
+      "put": {
```
Comment (majastrz) [inline specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json:241]: >put [](start = 7, length = 3)

Put on a resource collection is not allowed. See [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md).

File: specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json [hunk 3]
```diff
@@ -0,0 +1,866 @@
...
+        "id": {
+          "description": "The resource uri",
+          "type": "string"
+        },
+        "type": {
+          "readOnly": true,
+          "type": "string",
+          "description": "Resource type."
+        },
+        "tags": {
```
Comment (majastrz) [inline specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json:443]: >tags [](start = 9, length = 4)

You have the `Microsoft.SaaS/saasresources` resource type modeled as a tenant resource, which is a proxy a proxy resource, which doesn't support tags the same way tracked resources do.

File: specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json [hunk 4]
```diff
@@ -0,0 +1,866 @@
...
+          "type": "string",
+          "description": "Resource type."
+        },
+        "properties": {
+          "$ref": "#/definitions/SaasCreationProperties",
+          "description": "Properties of the SaaS resource that are relevant for creation."
+        }
+      }
+    },
+    "SaasProperties": {
```
Comment (majastrz) [inline specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json:472]: >SaasProperties [](start = 5, length = 14)

Are any of the properties read-only? If yes, they should be marked as such in the swagger.

File: specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json [hunk 5]
```diff
@@ -0,0 +1,866 @@
...
+          "type": "number",
+          "description": "The reasonCode"
+        },
+        "feedback": {
+          "type": "string",
+          "description": "the feedback"
+        }
+      }
+    },
+    "ErrorResponse": {
```
Comment (majastrz) [inline specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json:804]: >ErrorResponse [](start = 5, length = 13)

Please take a look at [https://github.com/Azure/azure-rest-api-specs/blob/master/specification/common-types/resource-management/v2/types.json](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/common-types/resource-management/v2/types.json). You should be able to reuse the error definition from that file. (Swagger allows you to reference content from another swagger file.)

File: specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json [hunk 6]
```diff
@@ -0,0 +1,866 @@
...
+      "properties": {
+        "nextLink": {
+          "description": "the next link to query to get the remaining results.",
+          "type": "string"
+        },
+        "value": {
+          "description": "the value of response.",
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/SaasAppOperation"
```
Comment (majastrz) [inline specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json:623]: >SaasAppOperation [](start = 35, length = 16)

Please take a look at [https://github.com/Azure/azure-rest-api-specs/blob/master/specification/common-types/resource-management/v2/types.json](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/common-types/resource-management/v2/types.json). You should be able to reuse the operation definition from that file. (Swagger allows you to reference content from another swagger file.)

File: specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json (hunk 2 above)
Comment (diyakobo) [reply specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json:241]: its a put on a single tenant level resource

File: specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json (hunk 3 above)
Comment (diyakobo) [reply specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json:443]: We use this property for Subscription level resources,
for tenant level resources its always null.

File: specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json (hunk 2 above)
Comment (majastrz) [reply specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json:241]: That is also compliant with the resource provider contract in ARM either as `saasresources` segment is in the place of the collection segment. You mentioned that this is an old API that you are just documenting. Can you confirm that this applies here as well?

---
In reply to: [529391536](https://github.com/Azure/azure-rest-api-specs/pull/11680#discussion_r529391536) [](ancestors = 529391536)

Comment (diyakobo) [reply specification/saas/resource-manager/Microsoft.SaaS/preview/2018-03-01-beta/saas.json:241]: yes, its an existing API that I'm documenting here, and that there are no changes in the service.

**Review by majastrz — APPROVED** [review]
I will sign off on this from ARM side based on the fact that you are only documenting an existing API. Some aspects of this API violate the RPC and ARM design guidelines, so you should not carry these approaches to new services/APIs.

**Review by weidongxu-microsoft — APPROVED** [review]

**Comment by weidongxu-microsoft** [comment]
Suppress on `DeleteMustNotHaveRequestBody` is approved by ARM.

In new api-version this should not be violated.

**Pull request merged**